tests/core/test_tracker_stores.py

```diff
@@ -976,3 +982,328 @@ def test_create_awaitable_tracker_store_with_endpoint_config():
+    assert tracker == tracker_with_restarted_event
+
+
+async def test_fail_safe_tracker_store_retrieve_full_tracker_with_exception(
+    caplog: LogCaptureFixture,
+) -> None:
+    primary_tracker_store = MagicMock()
+    primary_tracker_store.domain = Domain.empty()
+    primary_tracker_store.event_broker = None
+
+    exception = Exception("Something went wrong")
+    primary_tracker_store.retrieve_full_tracker = AsyncMock(side_effect=exception)
+
+    tracker_store = FailSafeTrackerStore(primary_tracker_store)
+    with caplog.at_level(logging.ERROR):
+        await tracker_store.retrieve_full_tracker("some_id")
+
+    assert "Error happened when trying to retrieve conversation tracker" in caplog.text
+    assert f"Please investigate the following error: {exception}." in caplog.text
+
+
+async def test_sql_get_or_create_full_tracker_without_action_listen() -> None:
+    tracker_store = SQLTrackerStore(Domain.empty())
+    sender_id = uuid.uuid4().hex
+    tracker = await tracker_store.get_or_create_full_tracker(
+        sender_id=sender_id, append_action_listen=False
+    )
+    assert tracker.sender_id == sender_id
+    assert tracker.events == deque()
+
+
+async def test_sql_get_or_create_full_tracker_with_action_listen() -> None:
+    tracker_store = SQLTrackerStore(Domain.empty())
+    sender_id = uuid.uuid4().hex
+    tracker = await tracker_store.get_or_create_full_tracker(
+        sender_id=sender_id, append_action_listen=True
+    )
+    assert tracker.sender_id == sender_id
+    assert tracker.events == deque([ActionExecuted(ACTION_LISTEN_NAME)])
+
+
+async def test_sql_get_or_create_full_tracker_with_existing_tracker(
+    tracker_with_restarted_event: DialogueStateTracker,
+) -> None:
+    sender_id = tracker_with_restarted_event.sender_id
+
+    tracker_store = SQLTrackerStore(Domain.empty())
+    await tracker_store.save(tracker_with_restarted_event)
+
+    tracker = await tracker_store.get_or_create_full_tracker(
+        sender_id=sender_id, append_action_listen=False
+    )
+    assert tracker.sender_id == sender_id
+    assert tracker.events == deque(tracker_with_restarted_event.events)
+
+
+async def test_sql_tracker_store_retrieve_full_tracker(
+    domain: Domain, tracker_with_restarted_event: DialogueStateTracker
+) -> None:
+    tracker_store = SQLTrackerStore(domain)
+    sender_id = tracker_with_restarted_event.sender_id
+    await tracker_store.save(tracker_with_restarted_event)
+
+    tracker = await tracker_store.retrieve_full_tracker(sender_id)
+    assert tracker == tracker_with_restarted_event
+
+
+async def test_sql_tracker_store_retrieve(
+    domain: Domain,
+    tracker_with_restarted_event: DialogueStateTracker,
+) -> None:
+    tracker_store = SQLTrackerStore(domain)
+    sender_id = tracker_with_restarted_event.sender_id
+    await tracker_store.save(tracker_with_restarted_event)
+
+    split_conversations = rasa.shared.core.events.split_events(
```

Comment:
Can we replace this with the raw list of items?
It would be easier to read what the expected result should be.

Reply:
Same for other occurrences of this line.